File: unity/Assets/Scripts/Multiplayer/Rooms/Updates/Updates.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using UnityEngine;

namespace Michitai.Multiplayer.Rooms.Updates
{
    /// <summary>
    /// Provides static methods for managing room updates.
    /// Handles sending updates to players and polling for received updates.
    /// Uses Unity's JsonUtility for serialization.
    /// </summary>
    public static class Updates
    {
        /// <summary>
        /// Sends updates to specific players in the room.
        /// </summary>
        /// <typeparam name="T">The type of update data.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="request">The update request containing targets and data.</param>
        /// <param name="ct">Cancellation token.</param>
        /// <returns>Task containing the update players response.</returns>
        public static Task<UpdatePlayersResponse> UpdatePlayersAsync<T>(Client client, string playerToken, UpdatePlayers<T> request, CancellationToken ct = default) where T : class, new()
            => client.Send<UpdatePlayersResponse>(HttpMethod.Post, client.Url(Endpoints.GameRoomUpdates, $"&player_token={playerToken}"),
                new UpdatePlayersRequest(request.TargetPlayers, request.Type, JsonUtility.ToJson(request.Data), request.TargetPlayersIds), ct);

        /// <summary>
        /// Polls for updates that were sent to the current player.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="request">The poll request containing filters.</param>
        /// <param name="lastUpdateId">Optional ID of the last update to poll from.</param>
        /// <param name="ct">Cancellation token.</param>
        /// <returns>Task containing the poll updates response with received updates.</returns>
        public static Task<PollUpdatesResponse> PollUpdatesAsync(Client client, string playerToken, PollUpdates request, string lastUpdateId = null, CancellationToken ct = default)
            => client.Send<PollUpdatesResponse>(HttpMethod.Post, client.Url(Endpoints.GameRoomUpdatesPoll, $"&player_token={playerToken}"),
                new PollUpdatesRequest(request.FromPlayers, request.FromPlayersIds, request.LastUpdate), ct);
    }
}

File: java/src/main/java/com/michitai/multiplayer/rooms/updates/Updates.java
package com.michitai.multiplayer.rooms.updates;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.michitai.multiplayer.Client;

import java.io.IOException;

/**
 * Provides methods for managing room updates.
 * Handles sending updates to players and polling for received updates.
 */
public class Updates {
    private static final ObjectMapper objectMapper = Client.JSON_MAPPER;

    /**
     * Sends updates to specific players in the room.
     *
     * @param client The API client instance.
     * @param playerToken The player's authentication token.
     * @param request The update request containing targets and data.
     * @return Response containing the update IDs and target player information.
     * @throws IOException if the request fails.
     */
    public static UpdatePlayersResponse updatePlayers(Client client, String playerToken, UpdatePlayers<?> request) throws IOException {
        String dataJson = request.getData() != null ? objectMapper.writeValueAsString(request.getData()) : null;
        UpdatePlayersRequest updateRequest = new UpdatePlayersRequest(
            request.getTargetPlayers(), 
            request.getType(), 
            dataJson, 
            request.getTargetPlayersIds()
        );
        return client.post(client.url(Endpoints.GAME_ROOM_UPDATES, "&player_token=" + playerToken), updateRequest, UpdatePlayersResponse.class);
    }

    /**
     * Polls for updates that were sent to the current player.
     *
     * @param client The API client instance.
     * @param playerToken The player's authentication token.
     * @param request The poll request containing filters.
     * @return Response containing the received updates.
     * @throws IOException if the request fails.
     */
    public static PollUpdatesResponse<?> pollUpdates(Client client, String playerToken, PollUpdates request) throws IOException {
        return pollUpdates(client, playerToken, request, Object.class);
    }

    /**
     * Polls for updates that were sent to the current player.
     *
     * @param <T> The type to deserialize update data into.
     * @param client The API client instance.
     * @param playerToken The player's authentication token.
     * @param request The poll request containing filters.
     * @param dataType The class to deserialize update data into.
     * @return Response containing the received updates with typed data.
     * @throws IOException if the request fails.
     */
    public static <T> PollUpdatesResponse<T> pollUpdates(Client client, String playerToken, PollUpdates request, Class<T> dataType) throws IOException {
        PollUpdatesRequest pollRequest = new PollUpdatesRequest(
            request.getFromPlayers(), 
            request.getFromPlayersIds(), 
            request.getLastUpdate()
        );
        return client.post(client.url(Endpoints.GAME_ROOM_UPDATES_POLL, "&player_token=" + playerToken), pollRequest,
            client.parametricType(PollUpdatesResponse.class, dataType));
    }
}
